### PR TITLE
[WPE][GTK] Symboilze `StackTrace` using `libbacktrace`

### PR DESCRIPTION
**Pull request by obyknovenius**
#### e9b9e04d516354cd67b372a4fe5ef7c3fd88b845
<pre>
[WPE][GTK] Symboilze `StackTrace` using `libbacktrace`
<a href="https://bugs.webkit.org/show_bug.cgi?id=262104">https://bugs.webkit.org/show_bug.cgi?id=262104</a>

Reviewed by Michael Catanzaro.

This patch adds a new `USE_LIBBACKTRACE` CMake option.
When enabled, `StackTraceSymbolResolver` will use `libbacktrace`[1]
to resolve addresses to function names.

[1] <a href="https://github.com/ianlancetaylor/libbacktrace">https://github.com/ianlancetaylor/libbacktrace</a>

* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/StackTrace.cpp:
(WTF::backtraceState):
(WTF::backtraceSyminfoCallback):
(WTF::backtraceFullCallback):
(WTF::symbolize):
* Source/WTF/wtf/StackTrace.h:
(WTF::StackTraceSymbolResolver::forEach const):
* Source/cmake/FindLibBacktrace.cmake: Added.
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/268579@main">https://commits.webkit.org/268579@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7d3685910ecb0e8511be4233e5ce5986addcc0b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21969 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18735 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23752 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20667 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20220 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20215 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17445 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22819 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17385 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18242 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24495 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/17451 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18462 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18418 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22485 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/19433 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16139 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/23474 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18207 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4818 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22550 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/24730 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18836 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5470 "Passed tests") |

### CODE REVIEW

File: Source/WTF/wtf/PlatformWPE.cmake [hunk 1]
```diff
@@ -64,3 +64,13 @@ list(APPEND WTF_SYSTEM_INCLUDE_DIRECTORIES
...
+
+if (USE_LIBBACKTRACE)
+    list(APPEND WTF_SYSTEM_INCLUDE_DIRECTORIES
+        ${LIBBACKTRACE_INCLUDE_DIRS}
+    )
+
+    list(APPEND WTF_LIBRARIES
+        ${LIBBACKTRACE_LIBRARIES}
+    )
+endif ()
```
Comment (mcatanzaro) [inline Source/WTF/wtf/PlatformWPE.cmake]: Please use the imported target `LIBBACKTRACE::LIBBACKTRACE` instead, matching how threads, journald, and zlib is handled above. What you've got here is the older way of doing things that just hasn't been cleaned up yet. Same for PlatformGTK.cmake.

File: Source/cmake/OptionsGTK.cmake [hunk 2]
```diff
@@ -61,6 +61,7 @@ WEBKIT_OPTION_DEFINE(USE_OPENGL_OR_ES "Whether to use OpenGL or ES." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_OPENJPEG "Whether to enable support for JPEG2000 images." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_SOUP2 "Whether to enable usage of Soup 2 instead of Soup 3." PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(USE_WOFF2 "Whether to enable support for WOFF2 Web Fonts." PUBLIC ON)
+WEBKIT_OPTION_DEFINE(USE_LIBBACKTRACE "Whether to enable usage of libbacktrace." PUBLIC OFF)
```
Comment (mcatanzaro) [inline Source/cmake/OptionsGTK.cmake]: Please alphabetize it and turn it ON by default. Same for OptionsWPE.cmake.

File: Source/cmake/OptionsGTK.cmake [hunk 3]
```diff
@@ -471,6 +472,10 @@ if (USE_LCMS)
     endif ()
 endif ()
 
+if (USE_LIBBACKTRACE)
+  find_package(LibBacktrace REQUIRED)
```
Comment (mcatanzaro) [inline Source/cmake/OptionsGTK.cmake]: Please follow the pattern for the other dependencies above and leave a nice error message: "libbacktrace is required for USE_LIBBACKTRACE."

I see Thunder is not following this pattern, but that's OK because it's hidden behind private options. Yours is public.

File: Source/WTF/wtf/StackTrace.cpp [hunk 4]
```diff
@@ -45,6 +45,55 @@ void WTFGetBacktrace(void** stack, int* size)
 
 namespace WTF {
 
+#if USE(LIBBACKTRACE)
+static struct backtrace_state* backtraceState()
+{
+    static struct backtrace_state* backtraceState = backtrace_create_state(nullptr, 1, nullptr, nullptr);
```
Comment (mcatanzaro) [inline Source/WTF/wtf/StackTrace.cpp]: Please use wtf/NeverDestroyed for this, to document your intention and ensure the destructor doesn't run in an exit handler.

File: Source/WTF/wtf/StackTrace.cpp [hunk 5]
```diff
@@ -45,6 +45,55 @@ void WTFGetBacktrace(void** stack, int* size)
...
+    for (int i = 0; i < size; ++i) {
+        uintptr_t pc = reinterpret_cast<uintptr_t>(addresses[i]);
+        const char* symbol;
+
+        backtrace_pcinfo(state, pc, backtraceFullCallback, nullptr, &symbol);
+        if (!symbol)
+            backtrace_syminfo(backtraceState(), pc, backtraceSyminfoCallback, nullptr, &symbol);
+
+        if (symbol) {
+            char* demangled = abi::__cxa_demangle(symbol, nullptr, nullptr, nullptr);
```
Comment (mcatanzaro) [inline Source/WTF/wtf/StackTrace.cpp:90]: I see this is a libstdc++ extension. Is it supported by libc++?

This seems sufficiently nonportable that normally I'd like to see a feature test for it, but in this case I guess it's probably OK because libbacktrace is inherently specific to GCC and Clang. That said, please do make sure it won't break with libc++.

File: Source/WTF/wtf/StackTrace.cpp [hunk 6]
```diff
@@ -45,6 +45,55 @@ void WTFGetBacktrace(void** stack, int* size)
...
+    return 0;
+}
+
+const char** symbolize(void* const* addresses, int size)
+{
+    struct backtrace_state* state = backtraceState();
+    if (!state)
+        return nullptr;
+
+    const char** symbols = (const char**)malloc(sizeof(const char*) * size);
```
Comment (mcatanzaro) [inline Source/WTF/wtf/StackTrace.cpp]: Style: use `static_cast` wherever possible and `reinterpret_cast` otherwise; only do a C-style cast when the type you're casting from is uncertain and could be either a pointer or an integer, which is basically just EGL types.

```suggestion
    const char** symbols = static_cast<const char**>(malloc(sizeof(const char*) * size));
```

File: Source/WTF/wtf/StackTrace.cpp [hunk 7]
```diff
@@ -45,6 +45,55 @@ void WTFGetBacktrace(void** stack, int* size)
...
+            backtrace_syminfo(backtraceState(), pc, backtraceSyminfoCallback, nullptr, &symbol);
+
+        if (symbol) {
+            char* demangled = abi::__cxa_demangle(symbol, nullptr, nullptr, nullptr);
+            if (demangled)
+                symbols[i] = demangled;
+            else
+                symbols[i] = symbol;
+        } else
+            symbols[i] = "???";
```
Comment (mcatanzaro) [inline Source/WTF/wtf/StackTrace.cpp]: Since the result of `abi::__cxa_demangle` needs to be freed, you'll need to ensure these strings are heap-allocated too. Normally we would use `g_strdup` for this, but since this code doesn't use GLib you'll want `strdup` from string.h:

```suggestion
                symbols[i] = strdup(symbol);
        } else
            symbols[i] = strdup("???");
```

File: Source/WTF/wtf/StackTrace.cpp [hunk 8]
```diff
@@ -45,6 +45,55 @@ void WTFGetBacktrace(void** stack, int* size)
...
+}
+
+static int backtraceFullCallback(void* data, uintptr_t, const char*, int, const char* function)
+{
+    const char** symbol = static_cast<const char**>(data);
+    *symbol = function;
+    return 0;
+}
+
+const char** symbolize(void* const* addresses, int size)
```
Comment (mcatanzaro) [inline Source/WTF/wtf/StackTrace.cpp]: I see you're matching a preexisting interface here, but returning memory via raw pointer is pretty bad because it's easy to mess up. In this case, I see the return value is "transfer container" in that the array of strings but not the strings themselves need to be freed, and it will be freed once using `free` at the bottom of `StackTraceSymbolResolver`. Alas, that means the symbols returned by `abi::__cxa_demangle` are leaked. So we really need to return `char**` rather than `const char**` and each string in the array will need to be freed as well.

A better solution would be to convert everything to use `UniquePtr` so the caller doesn't have to worry about how to free it.

**Review by mcatanzaro — CHANGES_REQUESTED** [review]
Nice. This is one the right track, but there is a memory leak...

File: Source/WTF/wtf/PlatformWPE.cmake (hunk 1 above)
Comment (obyknovenius) [reply Source/WTF/wtf/PlatformWPE.cmake]: Done.

File: Source/cmake/OptionsGTK.cmake (hunk 2 above)
Comment (obyknovenius) [reply Source/cmake/OptionsGTK.cmake]: I alphabetize it. I can not turn it ON for now because it will break the bots. `libbacktrace` is a separate library and it's not preinstalled yet. I'm going to Flatpak SDK and ghbuild to include it in follow-up PRs. When it's done we can turn the flat on by default.

File: Source/cmake/OptionsGTK.cmake (hunk 3 above)
Comment (obyknovenius) [reply Source/cmake/OptionsGTK.cmake]: Done.

File: Source/WTF/wtf/StackTrace.cpp (hunk 4 above)
Comment (obyknovenius) [reply Source/WTF/wtf/StackTrace.cpp]: Done.

File: Source/WTF/wtf/StackTrace.cpp (hunk 6 above)
Comment (obyknovenius) [reply Source/WTF/wtf/StackTrace.cpp]: Done.

File: Source/WTF/wtf/StackTrace.cpp (hunk 7 above)
Comment (obyknovenius) [reply Source/WTF/wtf/StackTrace.cpp]: You're right. Done.

File: Source/WTF/wtf/StackTrace.cpp (hunk 8 above)
Comment (obyknovenius) [reply Source/WTF/wtf/StackTrace.cpp]: You're right. It was a leak. I fixed it by freeing every string. Using `UniquePtr` makes the code much more verbose. I have an idea how we can refactor `StackTraceSymbolResolver` but I want to land this first.

File: Source/WTF/wtf/StackTrace.cpp (hunk 5 above)
Comment (obyknovenius) [reply Source/WTF/wtf/StackTrace.cpp:90]: I don't know how to test it with `libc++`. Any suggestions?

File: Source/cmake/OptionsGTK.cmake (hunk 2 above)
Comment (mcatanzaro) [reply Source/cmake/OptionsGTK.cmake]: OK, this is fine, just remember that the work is only useful once you manage to turn it ON, because hardly anybody will ever manually enable this option. (And if it's not turned ON sooner or later, we'll eventually remove it.)

File: Source/WTF/wtf/StackTrace.cpp (hunk 5 above)
Comment (mcatanzaro) [reply Source/WTF/wtf/StackTrace.cpp:90]: Try building with clang and use `-stdlib=libc++` in `CMAKE_C_FLAGS` and `CMAKE_CXX_FLAGS`. If it works then you're good.

Comment (mcatanzaro) [reply Source/WTF/wtf/StackTrace.cpp:90]: I searched through the [libc++ source code](https://github.com/llvm/llvm-project/tree/main/libcxx) and found a bunch of references to `__cxa_demangle`. It's implemented; you're fine.

File: Source/WTF/wtf/StackTrace.cpp (hunk 8 above)
Comment (mcatanzaro) [reply Source/WTF/wtf/StackTrace.cpp]: I would have expected `UniquePtr` to simplify the code, not complicate it. Anyway, this is fine.

File: Source/WTF/wtf/ListHashSet.h [hunk 9]
```diff
@@ -357,6 +357,9 @@ inline ListHashSet<T, U>::ListHashSet(std::initializer_list<T> initializerList)
 
 template<typename T, typename U>
 inline ListHashSet<T, U>::ListHashSet(const ListHashSet& other)
+#if CHECK_HASHTABLE_ITERATORS
+    : CanMakeWeakPtr<ListHashSet<T, U>, WeakPtrFactoryInitialization::Eager>(other)
+#endif
```
Comment (mcatanzaro) [inline Source/WTF/wtf/ListHashSet.h]: This looks like a mistake? Surely this doesn't belong in this pull request?

**Review by mcatanzaro — APPROVED** [review]
OK, good except for the strange change to ListHashSet.h.

File: Source/WTF/wtf/ListHashSet.h (hunk 9 above)
Comment (obyknovenius) [reply Source/WTF/wtf/ListHashSet.h]: I needed this to build the debug version for some reason. Removed it.
